Zonal stats fixes and improvements

nyalldawson opened the pull request:
This PR fixes a bunch of identified issues with the zonal stats code, and greatly speeds up the algorithm in the case of small polygons vs pixel sizes. Lots of new tests too!

nyalldawson committed:
Use unique_ptr

nyalldawson committed:
Use doubles, not float values, within zonal stats

nyalldawson committed:
Avoid use of raw GEOS API

nyalldawson committed:
Use prepared geometries to check for intersection between
pixel rectangles and polygons when doing exact statistics
calculations in zonal stats

Sppeds up zonal stats calculation by ~10x when pixel size
is large (compared to polygon sizes)

nyalldawson committed:
Minor refactoring, avoid use of old style cast

nyalldawson committed:
Ensure zonal stats respsects all user set no data values

nyalldawson committed:
Add thread safe method of constructing QgsZonalStatistics,
using a QgsRasterInterface instead of a QgsRasterLayer

nyalldawson committed:
Fix incorrect calculation of zonal statistics

Raster block extent was based on intersection of raster vs
feature's bounding box, which was not necessarily snapped to
multiples of the pixel size. But the pixel center point/extent
was being calculated as though the retrieved extent was an
exact multiple of the pixel size. This led to incorrect
retrieval of pixel values.

nyalldawson committed:
Fix zonal statistics calculations when pixel size is greater than
polygon size

Fixes #17159

nyalldawson committed:
Fix zonal stats min/max values when pixel area is greater than
polygon area

nyalldawson committed:
Make zonal statistics algorithm thread safe

nyalldawson committed:
Update processing test results

nyalldawson commented:
@nirvn the zonal histogram algorithm suffers from many of the same bugs - it would be nice to remove the duplicate code here :)

nyalldawson commented:
@nirvn you should be able to make those functions take a function argument, and pass a lambda function in which handles the calculation of the result

nyalldawson commented:
@nirvn those methods aren't exposed publically. What I'd suggest doing is making a QgsZonalStatisticsUtils class, with the three methods as static members, and the whole class not exposed to python. Much like how the overlay utils is done.

nirvn commented:
@nyalldawson , OK, I'll undertake that when this PR is merged.

nirvn commented on src/analysis/vector/qgszonalstatistics.cpp:
```diff
@@ -439,59 +430,61 @@ void QgsZonalStatistics::statisticsFromMiddlePointTest( const QgsGeometry &poly,
+bool QgsZonalStatistics::validPixel( double value ) const
 {
-  return !( value == mInputNodataValue || std::isnan( value ) );
+  return !std::isnan( value );
```
Do we really need this validPixel function here?

nyalldawson replied:
Do you mean "do we need the check" or "does this have to be a function"?

nirvn replied:
@nyalldawson , "does this have to be a function?" :smile:

nyalldawson replied:
Not sure... I'm thinking there's *probably* an issue if non-finite pixel values are present here which should also be caught by the function, but I'm reluctant to add this extra check without knowing that there is actually an issue. So I'd keep it like this to make it easy to possibly refine like this in future....

nyalldawson replied:
(it's also called in more than one place)

nirvn replied:
OK

pblottiere commented on src/analysis/vector/qgszonalstatistics.cpp:
```diff
@@ -31,12 +31,37 @@
+  if ( mCellSizeX < 0 )
+  {
+    mCellSizeX = -mCellSizeX;
+  }
```
Would it be possible to do something like `, mCellSizeX( std::abs( rasterUnitsPerPixelX ) )` instead?

(same question about `mCellSizeY`)

nyalldawson replied:
Fixed

pblottiere commented on src/analysis/vector/qgszonalstatistics.h:
```diff
@@ -106,45 +140,48 @@ class ANALYSIS_EXPORT QgsZonalStatistics
+        double sum = 0.0;
+        double count = 0.0;
+        double max = std::numeric_limits<double>::lowest();
+        double min = std::numeric_limits<double>::max();
```
I didn't know about initializing double/float members in headers...

nyalldawson replied:
It's a C++11 feature, along with the nice delegated constructor I've used here :)

pblottiere replied:
Is it something we want to generalize? I mean, I was aware that we wanted to initialize nullptr and boolean in headers. But do we want to initialize all members directly in headers if possible? 

Just to know the rules for the future :).

And yes, delegate constructors rocks :+1:! But last time I tried, it seems to me that I had some issues for compiling on Windows... However, it was some time ago :).

nyalldawson replied:
In my view, yes initialise in headers wherever possible. It reduces the chance of uninitialised members being introduced when code is refactored, and reduces duplication (default values explicitly included in all individual constructors). 

I've run the associated clang tidy modernize fixits over the whole code previously, but they haven't caught everything (I think the fixit fails when there's multiple constructors, and it also doesn't like initialisation in the body of constructors (which some old old qgis code does)). So if you see this anywhere, please fix :)

pblottiere replied:
OK, thanks for the explanation Nyall! I'll pay attention to these points :)

nyalldawson committed:
Better constructor, dox

nyalldawson committed:
Prevent background execution of zonal stats

This algorithm is NEVER safe to run in the background, because
it directly modifies a layer in place.

We should:
- make a variant of this algorithm which returns a new layer
AND
- refactor the QgsZonalStats code to create a separate map
of feature->stats, and then modify the existing Zonal Stats
algorithm to apply these changes in a post processing step

nyalldawson committed:
Use a raster iterator for zonal stats, to optimise zonal stats
calculation of large zones with lots of corresponding raster pixels
